### PR TITLE
Opt boxing

### PR DESCRIPTION
**Pull request by homuroll**

### CODE REVIEW

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/boxing/IrBoxCounterField.kt [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+        fun get(context: Context, suffix: String = "") = counters.getOrPut(context) { IrBoxCounterField(context, suffix) }
+    }
+}
+
+internal fun IrBuilderWithScope.irInc(counter: IrBoxCounterField): IrSetFieldImpl {
+    val context = counter.context
+    val stdlib = context.irModules.filterKeys { it.endsWith("stdlib") }.values.firstOrNull()
+            ?: context.ir.irModule
+    val intClass = stdlib.files
```
Comment (homuroll) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/boxing/IrBoxCounterField.kt]: `this.context.irBuiltIns.intClass`

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/boxing/IrBoxCounterField.kt [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            .first { it.fqName.toString() == "kotlin" && it.fileEntry.name.endsWith("Primitives.kt") }
+            .declarations
+            .first { it.nameForIrSerialization.toString() == "Int" } as IrClass
+    val increment = intClass.declarations.first { it.nameForIrSerialization.toString() == "inc" } as IrFunction
+
+    return irSetField(null, counter.field, irCall(increment).also { it.dispatchReceiver = irGetField(null, counter.field) })
+}
+
+internal fun IrField.isBoxingCounter() = name.toString() == "${"$"}boxingCounter"
```
Comment (homuroll) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/boxing/IrBoxCounterField.kt]: `"\$boxingCounter"`

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/boxing/TypeParameterEliminator.kt [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        val symbol = getTypeArgument(0)?.classifierOrNull
+        return symbol in typeParameterMapping && typeParameterMapping[symbol]?.isPrimitiveType() == true
+    }
+}
+
+/**
+ * Copy of [DeepCopyTypeRemapper] with additional remapping for eliminated type parameters.
+ * TODO provide delegate
+ */
+class TypeRemapperWithParametersElimination(
```
Comment (homuroll) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/boxing/TypeParameterEliminator.kt]: Take a look at `DeepCopyIrTreeWithDescriptors.kt` in Kotlin repository. There is class `InlinerTypeRemapper` over there, it does similar work.

**Comment by SvyatoslavScherbina** [comment]
Abandoned, closing.

Also we are migrating the development to https://github.com/JetBrains/kotlin, so keeping the PR open in the old repository doesn't make much sense.

**Pull request closed without merging**